### PR TITLE
cmake: use the imported target of FindOpenSSL module

### PR DESCRIPTION
**Pull request by jimmy-park**
I'm trying to [build OpenSSL in CMake](https://github.com/jimmy-park/openssl-cmake) and link it to libssh2. However libssh2 requires some OpenSSL headers that aren't generated yet. We can avoid this by the target dependency. If we use the imported target from [FindOpenSSL module](https://cmake.org/cmake/help/latest/module/FindOpenSSL.html) instead of `OPENSSL_LIBRARIES` which is a file path, libssh2 depends OpenSSL and could be built after OpenSSL.

Here is the minimal example.

```cmake
cmake_minimum_required(VERSION 3.25)

project(ssh LANGUAGES CXX)

include(FetchContent)

FetchContent_Declare(
    openssl-cmake
    URL https://github.com/jimmy-park/openssl-cmake/archive/3.2.1.tar.gz
)
FetchContent_MakeAvailable(openssl-cmake)

FetchContent_Declare(
    libssh2
    URL https://github.com/jimmy-park/libssh2/archive/build-with-openssl.tar.gz

    # Failed
    # URL https://github.com/libssh2/libssh2/archive/master.tar.gz
)
FetchContent_MakeAvailable(libssh2)

add_executable(main)
target_sources(main PRIVATE main.cpp)
target_link_libraries(main PRIVATE libssh2::libssh2)
```

And since the `runner` target in tests also requires libssh2 with OpenSSL, we can build it after libssh2 using `target_link_libraries()`.

### CODE REVIEW

**Comment by vszakats** [comment]
Thanks! Our CMake integration test fell over this: https://github.com/libssh2/libssh2/actions/runs/8050664975/job/22413249455?pr=1322#step:3:1808

File: cmake/libssh2-config.cmake.in [hunk 1]
```diff
@@ -1,6 +1,16 @@
...
 # SPDX-License-Identifier: BSD-3-Clause
 
+set(CRYPTO_BACKEND_LIST OpenSSL wolfSSL Libgcrypt mbedTLS)
+
+if("@CRYPTO_BACKEND@" IN_LIST CRYPTO_BACKEND_LIST)
+    find_package(@CRYPTO_BACKEND@)
+endif()
+
+if(@ZLIB_FOUND@)
+    find_package(ZLIB)
```
Comment (vszakats) [inline cmake/libssh2-config.cmake.in:11]:
```suggestion
  find_package(@CRYPTO_BACKEND@)
endif()

if(@ZLIB_FOUND@)
  find_package(ZLIB)
```

Would this work with wolfSSL, which uses lowercase in the root CMakeLists.txt `find_package`? Also would it work with `WinCNG` which doesn't use `find_package`? Would these use our own find logic stored under `./cmake`? 

TL;DR: is there a solution which doesn't need duplicating dependency detection in the distributed cmake config?

Comment (jimmy-park) [reply cmake/libssh2-config.cmake.in:11]: > Would this work with wolfSSL, which uses lowercase in the root CMakeLists.txt find_package? 

It will fail. I didn't notice that wolfSSL's package name and `CRYPTO_BACKEND` were different. It seems like each `CRYPTO_BACKEND` should be handled separately.

> Also would it work with WinCNG which doesn't use find_package?

WinCNG isn't included in `CRYPTO_BACKEND_LIST` for that reason.

> edit: in comparison, curl has this for OpenSSL and zlib only:
https://github.com/curl/curl/blob/39173f66e541db909069a4ce30d7590b76041596/CMake/curl-config.cmake.in#L26-L32
It's unclear why these two dependencies need this and not others, but we may go with that too if they are special. In a config file find_dependency seems to preferred over find_package?

Thanks for resource. `find_dependency()` would be more appropriate in this case. I'll try this way.

Comment (jimmy-park) [reply cmake/libssh2-config.cmake.in:11]: > TL;DR: is there a solution which doesn't need duplicating dependency detection in the distributed cmake config?

I think this is the proper way to handle package dependencies in CMake.
https://cmake.org/cmake/help/latest/guide/importing-exporting/index.html#:~:text=All%20required%20dependencies,the%20Stats%20package.

File: cmake/libssh2-config.cmake.in [hunk 2]
```diff
@@ -1,6 +1,29 @@
...
+  find_package(wolfssl)
+endif()
+
+if("@CRYPTO_BACKEND@" STREQUAL "Libgcrypt")
+  find_package(Libgcrypt)
+endif()
+
+if("@CRYPTO_BACKEND@" STREQUAL "mbedTLS")
+  find_package(mbedTLS)
+endif()
```
Comment (vszakats) [inline cmake/libssh2-config.cmake.in:21]: Thanks! Looking good.
I'm wondering if using `elseif()` would make this look sweeter, what do you think?

Comment (jimmy-park) [reply cmake/libssh2-config.cmake.in:21]: I also think `elsif()` would be nice. Wait a moment.

**Review by vszakats — APPROVED** [review]

**Comment by vszakats** [comment]
Thank you!
